micronota/db/tigrfam.py

```diff
@@ -0,0 +1,266 @@
+import tarfile
+from tempfile import mkdtemp
+from os.path import join, exists
+from ftplib import FTP
+from sqlite3 import connect
+
+from ..bfillings.hmmer import hmmpress_hmm
+
+
+def prepare_db(out_d, prefix='tigrfam_v15.0', force=False,
+               server='ftp.tigr.org',
+               path='pub/data/TIGRFAMs',
+               hmm='TIGRFAMs_15.0_HMM.LIB.gz',
+               metadata='TIGRFAMs_15.0_INFO.tar.gz'):
+    '''Download and prepare TIGRFAM database.
+
+    Parameters
+    ----------
+    out_d : str
+        The directory of output files.
+    prefix : str
+        The file name (without extensions) of the output files.
+    force : boolean
+        Whether to overwrite existing files
+    server : str
+        FTP server to download TIGRFAM files
+    path : str
+        The path on FTP to find files to download
+    hmm : str
+        The file name of hmm models
+    metadata : str
+        The file name of the metadata for the hmm models
+    '''
+
+    hmm_out = join(out_d, '%s.hmm' % prefix)
+    if exists(hmm_out) and not force:
+        print('Database %s already exists. Skip it.' % prefix)
+        return
+
+    metadata_out = join(out_d, '%s.db' % prefix)
+
+    try:
+        temp_dir = mkdtemp()
+        hmm_tmp = join(temp_dir, hmm)
+        metadata_tmp = join(temp_dir, metadata)
+        with FTP(server) as ftp:
+            ftp.login()
+            ftp.cwd(path)
+
+            # fetch metadata file
+            ftp.retrbinary('RETR %s' % metadata,
+                           open(metadata_tmp, 'wb').write)
+            with tarfile.open(metadata_tmp) as tar:
+                tar.extractall(temp_dir)
+                prepare_metadata(temp_dir, metadata_out)
+
+            # fetch HMM model file
+            ftp.retrbinary('RETR %s' % hmm,
+                           open(hmm_tmp, 'wb').write)
+            # gunzip and move the file
+            with gzip.open(hmm_tmp, 'rb') as i_f, open(hmm_out, 'wb') as o_f:
+                shutil.copyfileobj(i_f, o_f)
+            # don't forget to compress the hmm model file
+            hmmpress_hmm(hmm_out)
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+def prepare_metadata(in_d, fp):
+    '''Compile the metadata into sqlite3 database.
+
+    Parameters
+    ----------
+    in_d : str
+        The input directory containing XXX.INFO files.
+    fp : str
+        The output file path of sqlite3 database.
+
+    Notes
+    -----
+    The schema of the database file contains one table named `tigrfam` that
+    has following columns:
+
+    1. ``id``. TEXT. TIGRFAM ID.
+
+    2. ``tag``. TEXT. 2-letter tag described in the table above in this module.
+
+    3. ``value``. BLOB. The value of the tag.
+
+    4. ``transfer``. INTEGER. Used as the boolean. ``1`` means the ``value``
+       should be transferred to the query sequences as its annotation;
+       ``0`` means not.
+    '''
+    with connect(fp) as conn:
+        conn.execute("DROP TABLE IF EXISTS tigrfam")
```

Comment:
yes. that's the intention.

On Sun, Jan 31, 2016 at 8:31 PM, Yoshiki Vázquez Baeza <
notifications@github.com> wrote:

> In micronota/db/tigrfam.py
> <https://github.com/biocore/micronota/pull/40#discussion_r51378484>:
>
> > +    -----
> > +    The schema of the database file contains one table named `tigrfam` that
> > +    has following columns:
> > +
> > +    1. ``id``. TEXT. TIGRFAM ID.
> > +
> > +    2. ``tag``. TEXT. 2-letter tag described in the table above in this module.
> > +
> > +    3. ``value``. BLOB. The value of the tag.
> > +
> > +    4. ``transfer``. INTEGER. Used as the boolean. ``1`` means the ``value``
> > +       should be transferred to the query sequences as its annotation;
> > +       ``0`` means not.
> > +    '''
> > +    with connect(fp) as conn:
> > +        conn.execute("DROP TABLE IF EXISTS tigrfam")
>
> Just in case it would be worth noting that if the table already exists, it
> will be dropped?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/micronota/pull/40/files#r51378484>.
>